cranelift/codegen/src/isa/x64/encoding/evex.rs

```diff
@@ -131,17 +135,40 @@ impl EvexInstruction {
         self
     }
 
-    /// Set the register to use for the `rm` bits; many instructions use this as the "read from
-    /// register/memory" operand. Currently this does not support memory addressing (TODO).Setting
-    /// this affects both the ModRM byte (`rm` section) and the EVEX prefix (the extension bits for
-    /// register encodings > 8).
+    /// Set the register to use for the `rm` bits; many instructions use this
+    /// as the "read from register/memory" operand. Setting this affects both
+    /// the ModRM byte (`rm` section) and the EVEX prefix (the extension bits
+    /// for register encodings > 8).
     #[inline(always)]
-    pub fn rm(mut self, reg: impl Into<Register>) -> Self {
+    pub fn rm(mut self, reg: impl Into<RegisterOrAmode>) -> Self {
+        // NB: See Table 2-31. 32-Register Support in 64-bit Mode Using EVEX
+        // with Embedded REX Bits
         self.rm = reg.into();
-        let b = !(self.rm.0 >> 3) & 1;
-        let x = !(self.rm.0 >> 4) & 1;
-        self.write(Self::X, x as u32);
-        self.write(Self::B, b as u32);
+        let x = match &self.rm {
+            RegisterOrAmode::Register(r) => r.0 >> 4,
+            RegisterOrAmode::Amode(Amode::ImmRegRegShift { index, .. }) => {
+                index.to_real_reg().unwrap().hw_enc() >> 3
+            }
+
+            // These two modes technically don't use the X bit, so leave it at
+            // 0.
+            RegisterOrAmode::Amode(Amode::ImmReg { .. }) => 0,
+            RegisterOrAmode::Amode(Amode::RipRelative { .. }) => 0,
+        };
+        self.write(Self::X, u32::from(!x & 1));
+
+        let b = match &self.rm {
+            RegisterOrAmode::Register(r) => r.0 >> 3,
+            RegisterOrAmode::Amode(Amode::ImmReg { base, .. }) => {
+                base.to_real_reg().unwrap().hw_enc() >> 3
+            }
+            RegisterOrAmode::Amode(Amode::ImmRegRegShift { base, .. }) => {
+                base.to_real_reg().unwrap().hw_enc() >> 3
+            }
+            // The 4th bit of %rip is 0
+            RegisterOrAmode::Amode(Amode::RipRelative { .. }) => 0,
+        };
+        self.write(Self::B, u32::from(!b & 1));
```

Comment:
It took me a second to review this because I was worried the inversion would be double-applied or something like that. For reference for others in the future, `EVEX.RXB` are stored in inverted format (do we want to note that here?). I eventually conclude this is actually fine since the only inversions performed are `!b` and `!x`.